econml/dml/dynamic_dml.py

```diff
@@ -0,0 +1,789 @@
+        wrapped_final_model = _DynamicFinalWrapper(
+            StatsModelsLinearRegression(fit_intercept=False),
+            fit_cate_intercept=self.fit_cate_intercept,
+            featurizer=self.featurizer,
+            use_weight_trick=False)
+        return _LinearDynamicModelFinal(wrapped_final_model, n_periods=n_periods)
+
+    def _prefit(self, Y, T, *args, groups=None, only_final=False, **kwargs):
+        u_periods = np.unique(np.unique(groups, return_counts=True)[1])
+        if len(u_periods) > 1:
+            raise AttributeError(
+                "Imbalanced panel. Method currently expects only panels with equal number of periods. Pad your data")
+        self._n_periods = u_periods[0]
+        # generate an instance of the final model
+        self._ortho_learner_model_final = self._gen_ortho_learner_model_final(self._n_periods)
+        if not only_final:
+            # generate an instance of the nuisance model
+            self._ortho_learner_model_nuisance = self._gen_ortho_learner_model_nuisance(self._n_periods)
+        TreatmentExpansionMixin._prefit(self, Y, T, *args, **kwargs)
+
+    def _postfit(self, Y, T, *args, **kwargs):
+        super()._postfit(Y, T, *args, **kwargs)
+        # Set _d_t to effective number of treatments
+        self._d_t = (self._n_periods * self._d_t[0], ) if self._d_t else (self._n_periods, )
+
+    def _strata(self, Y, T, X=None, W=None, Z=None,
+                sample_weight=None, sample_var=None, groups=None,
+                cache_values=False, only_final=False, check_input=True):
+        # Required for bootstrap inference
+        return groups
+
+    @_deprecate_positional("X, and should be passed by keyword only. In a future release "
+                           "we will disallow passing X and W by position.", ['X', 'W'])
+    def fit(self, Y, T, X=None, W=None, *, sample_weight=None, sample_var=None, groups,
+            cache_values=False, inference=None):
+        """Estimate the counterfactual model from data, i.e. estimates function :math:`\\theta(\\cdot)`.
+
+        The input data must contain groups with the same size corresponding to the number
+        of time periods the treatments were assigned over.
+
+        The data should be preferably in panel format, with groups clustered together.
+        If group members do not appear together, the following is assumed:
+
+        * the first instance of a group in the dataset is assumed to correspond to the first period of that group
+        * the second instance of a group in the dataset is assumed to correspond to the
+          second period of that group
+
+        ...etc.
+
+        Parameters
+        ----------
+        Y: (n, d_y) matrix or vector of length n
+            Outcomes for each sample (required: n = n_groups * n_periods)
+        T: (n, d_t) matrix or vector of length n
+            Treatments for each sample (required: n = n_groups * n_periods)
+        X: optional(n, d_x) matrix or None (Default=None)
```

Comment:
It might be good to mention here and in the beginning of the docstring that heterogeneity is measured only with respect to the period[0] value of X and the rest of the X's are just used as time-varying controls. Since this is a complex part of the API.

Reply:
Done